src/core/lib/channel/promise_based_filter.h

```diff
@@ -380,6 +382,276 @@ class CallData<ChannelFilter, true> : public BaseCallData {
+      grpc_transport_stream_op_batch_finish_with_failure(
+          absl::exchange(send_trailing_metadata_batch_, nullptr),
+          GRPC_ERROR_REF(cancelled_error_), call_combiner());
+    }
+    send_trailing_state_ = SendTrailingState::kCancelled;
+  }
+
+  // Construct a promise that will "call" the next filter.
+  // Effectively:
+  //   - put the modified initial metadata into the batch being sent up.
+  //   - return a wrapper around PollTrailingMetadata as the promise.
+  ArenaPromise<TrailingMetadata> MakeNextPromise(
+      ClientInitialMetadata initial_metadata) {
+    GPR_ASSERT(recv_initial_state_ == RecvInitialState::kComplete);
+    GPR_ASSERT(UnwrapMetadata(std::move(initial_metadata)) ==
+               recv_initial_metadata_);
+    forward_recv_initial_metadata_callback_ = true;
+    return ArenaPromise<TrailingMetadata>(
+        [this]() { return PollTrailingMetadata(); });
+  }
+
+  // Wrapper to make it look like we're calling the next filter as a promise.
+  // All polls: await sending the trailing metadata, then foward it down the
+  // stack.
+  Poll<TrailingMetadata> PollTrailingMetadata() {
+    switch (send_trailing_state_) {
+      case SendTrailingState::kInitial:
+        return Pending{};
+      case SendTrailingState::kQueued:
+        return WrapMetadata(
+            send_trailing_metadata_batch_->payload->send_trailing_metadata
+                .send_trailing_metadata);
+      case SendTrailingState::kForwarded:
+        abort();  // unreachable
+      case SendTrailingState::kCancelled:
+        // We could translate cancelled_error to metadata and return it... BUT
+        // we're not gonna be running much longer and the results going to be
+        // ignored.
+        return Pending{};
+    }
+    GPR_UNREACHABLE_CODE(return Pending{});
+  }
+
+  static void RecvInitialMetadataReadyCallback(void* arg,
+                                               grpc_error_handle error) {
+    static_cast<CallData*>(arg)->RecvInitialMetadataReady(error);
+  }
+
+  void RecvInitialMetadataReady(grpc_error_handle error) {
+    GPR_ASSERT(recv_initial_state_ == RecvInitialState::kForwarded);
+    // If there was an error we just propagate that through
+    if (error != GRPC_ERROR_NONE) {
+      recv_initial_state_ = RecvInitialState::kResponded;
+      Closure::Run(DEBUG_LOCATION, original_recv_initial_metadata_ready_,
+                   GRPC_ERROR_REF(error));
+      return;
+    }
+    // Record that we've got the callback.
+    recv_initial_state_ = RecvInitialState::kComplete;
+
+    // Start the promise.
+    ScopedContext context(this);
+    // Construct the promise.
+    ChannelFilter* filter = static_cast<ChannelFilter*>(elem()->channel_data);
+    promise_ = filter->MakeCallPromise(
+        WrapMetadata(recv_initial_metadata_),
+        [this](ClientInitialMetadata initial_metadata) {
+          return MakeNextPromise(std::move(initial_metadata));
+        });
+    // Poll once.
+    bool own_error = false;
```

Comment:
This is a nice mechanism for preserving the error for the callback.  Can we use this same approach on the client side for the `recv_trailing_metadata_ready` callback?